multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,327 @@
+	var err error
+	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			return r.onPodDelete(ctx, req.NamespacedName.String(), true)
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	normalizedLabel := getNormalizedLabel(ns.Labels, pod.Labels, ns.Name)
+	return r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), normalizedLabel)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(predicate.LabelChangedPredicate{}).
+		Watches(&source.Kind{Type: &v1.Namespace{}}, handler.EnqueueRequestsFromMapFunc(r.namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// onPodDelete removes the Pod and label identity mapping from the cache, and deletes label identity
+// ResourceExport if necessary (the Pod deletion event causes a label identity no longer present in
+// the cluster).
+// Note that if releaseLabelMutex is set to false, r.labelMutex will continue to be held after the
+// function returns, which is useful for eliminating additional context switching when handling Pod
+// update events, since after deleting stale label the labelMutex will immediately need to be held
+// again for checking label add.
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string, releaseLabelMutex bool) (ctrl.Result, error) {
+	r.labelMutex.Lock()
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if requeue := r.requeueIfLabelResExportUpdateInProgress(originalLabel); requeue {
+			r.labelMutex.Unlock()
+			return ctrl.Result{Requeue: true}, nil
+		}
```

Comment:
Save the label to labelDeletesInProgress.

Reply:
I considered the problem you brought up yesterday and feel we don't need to save label to `labelExportUpdatesInProgress` here. This `labelExportUpdatesInProgress` is protected by labelMutex for the following sequences

1. lock -> check-in-progress ->  fail ->  return  -> unlock  
2. lock  -> check-in-progress  -> success  -> no-resource-crud-needed -> update-cache -> return -> unlock
3. lock  -> check-in-progress  -> success -> resource-crud-needed -> set-in-progress -> unlock -> api-call -> lock -> unset-in-progress -> update-cache -> return -> unlock

If you look at 1 and 3 in particular, there's no way two workers can pass the check-in-progress at the same time if they are trying to update resource for the same label. Hence, if  check-in-progress fails, we can simply requeue.